systems/trajectory_optimization/hybrid_dircon.cc

```diff
@@ -70,39 +73,44 @@ HybridDircon<T>::HybridDircon(
     }
 
     // initialize constraint lengths
+    num_kinematic_constraints_wo_skipping_.push_back(constraints_[i]->countConstraintsWithoutSkipping());
     num_kinematic_constraints_.push_back(constraints_[i]->countConstraints());
 
     // initialize decision variables
-    force_vars_.push_back(NewContinuousVariables(constraints_[i]->countConstraints() * num_time_samples[i], "lambda[" + std::to_string(i) + "]"));
-    collocation_force_vars_.push_back(NewContinuousVariables(constraints_[i]->countConstraints() * (num_time_samples[i] - 1), "lambda_c[" + std::to_string(i) + "]"));
-    collocation_slack_vars_.push_back(NewContinuousVariables(constraints_[i]->countConstraints() * (num_time_samples[i] - 1), "v_c[" + std::to_string(i) + "]"));
+    force_vars_.push_back(NewContinuousVariables(constraints_[i]->countConstraintsWithoutSkipping() * num_time_samples[i], "lambda[" + std::to_string(i) + "]"));
+    collocation_force_vars_.push_back(NewContinuousVariables(constraints_[i]->countConstraintsWithoutSkipping() * (num_time_samples[i] - 1), "lambda_c[" + std::to_string(i) + "]"));
+    collocation_slack_vars_.push_back(NewContinuousVariables(constraints_[i]->countConstraintsWithoutSkipping() * (num_time_samples[i] - 1), "v_c[" + std::to_string(i) + "]"));
     offset_vars_.push_back(NewContinuousVariables(options[i].getNumRelative(), "offset[" + std::to_string(i) + "]"));
     if (i > 0) {
-      impulse_vars_.push_back(NewContinuousVariables(constraints_[i]->countConstraints(), "impulse[" + std::to_string(i) + "]"));
+      impulse_vars_.push_back(NewContinuousVariables(constraints_[i]->countConstraintsWithoutSkipping(), "impulse[" + std::to_string(i) + "]"));
     }
 
-    auto constraint = std::make_shared<DirconDynamicConstraint<T>>(plant_, *constraints_[i]);
-
-    DRAKE_ASSERT(static_cast<int>(constraint->num_constraints()) == num_states());
+    // slack variable for unit norm constraint of quaternion floating-base
+    int num_quat_slack = (multibody::isQuaternion(plant))? 1 : 0;
 
     // For N-1 timesteps, add a constraint which depends on the knot
     // value along with the state and input vectors at that knot and the
     // next.
 
     //Adding dynamic constraints
+    auto constraint = std::make_shared<DirconDynamicConstraint<T>>(plant_, *constraints_[i], num_quat_slack);
```

Comment:
Shouldn't the quaternion constraint be created and added here, rather than in the main file?

Reply:
Yes. I made this change in the new commits